### PR TITLE
chips: apollo3: Support the flash controller 

### PR DESCRIPTION
**Pull request by alistair23**
### Pull Request Overview

The Apollo3 flash controller is not documented in the datasheet and not
even documented in the vendor HAL. Instead the vendor HAL calls to
special ROM functions to do flash read, write and erase operations. This
is to allow operating on flash while also executing from it.

This commit adds support for the flash by calling the same ROM
functions. There doesn't appear to be any other way to write or erase
flash.

### Testing Strategy

libtock-c K/V examples

### Documentation Updated

- [X] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [X] Ran `make prepush`.

### CODE REVIEW

**Comment by hudson-ayers** [comment]
Does every apollo3 have the same ROM contents? Is there a chance different versions of the chip would have these functions at different addresses?

**Comment by alistair23** [comment]
> Does every apollo3 have the same ROM contents? Is there a chance different versions of the chip would have these functions at different addresses?

The HAL uses a fixed address for all Apollo3 boards, so I assume they all have the same contents

File: chips/apollo3/src/flash_ctrl.rs [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+            return Err((ErrorCode::BUSY, buf));
+        }
+
+        let addr = (page_number * PAGE_SIZE) as u32;
+        let addr_ptr = addr as *mut u32;
+
+        // Call `int flash_program_main(uint32_t, uint32_t *, uint32_t *, uint32_t)` in the ROM code.
+        let ptr = 0x08000055 as *const ();
+        let f: extern "C" fn(u32, *mut u32, *mut u32, u32) -> c_int =
+            unsafe { core::mem::transmute(ptr) };
```
Comment (lschuermann) [inline chips/apollo3/src/flash_ctrl.rs:160]: I have not conducted an in-depth review yet, but this is a type of unsafe code that we don't yet have in Tock. Regardless of a deeper discussion about calling into ROM code, this should definitely be a cast to an unsafe function:

```suggestion
        // Call `int flash_program_main(uint32_t, uint32_t *, uint32_t *, uint32_t)` in the ROM code.
        let rom_flash_program_main: extern "C" fn(u32, *mut u32, *mut u32, u32) -> c_int =
            unsafe { core::mem::transmute(0x08000055 as *const ()) };
```

We'll also want to avoid having variable names like `ptr` and `f` in scope, _especially_ in such safety-critical code.

**Comment by lschuermann** [comment]
tl;dr: calling into ROM code seems fine for this particular case, but we'll want to avoid burying the transmute deep within the driver code. Splitting it out into a top-level unsafe wrapper function at the top of the file will make this more obvious, similar to this: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=87e76ddf802b2c7c507714fca3901efe

Thinking about calling into ROM code more, it seems that this is virtually unavoidable in this case. While we should be careful of this becoming precedent, this usage seems generally acceptable for a few reasons:
- There is no viable alternative, and in particular no specification of the hardware flash controller. It's unreasonable for us to reverse-engineer this flash controller. We're mimicking vendor firmware behavior here.
- This is limited to a particular chip. Any fallout from this ROM code will be limited to apollo3 boards, and this change does not change core kernel interfaces.
- Any function that provides unrestricted access to a flash controller is necessarily unsafe.

That being said, for posterity's sake, there's some unfortunate side effects of this:
- We cannot reason about all safety invariants on this function. We do not have access to the source code, so we could not, for instance, reason about whether this ROM code touches other global state, requires any form of initialization, or is safe to call in threaded (interrupt) contexts.
  - This makes it hard or impossible to reason about "safe wrappers" around these interfaces.
- We have no way to check or enforce that the ROM function we are calling into is the same between chips / revisions, etc. This will make issues arising from these calls hard to diagnose.

One issue with the way the code is currently structured is that this call is not clearly visible (like a regular `extern "C"` symbol definition is), but instead buried deep in driver code. This makes sense -- we can't easily define a symbol like `extern "C" { fn apollo3_rom_flash_erase_page(...) }` without then also playing linker tricks to have this symbol point to a particular address. I was hopeful that we could alleviate this by moving the function reference into a static, but Rust refuses to transmute an immediate (or even another static) into a function: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=6ab624f924deaa283391ad531e39ecdb

This is the best I could come up with is just splitting this out into another unsafe wrapper function:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=87e76ddf802b2c7c507714fca3901efe

**Comment by alistair23** [comment]
Transmute is now much more obvious

File: boards/apollo3/lora_things_plus/src/main.rs [hunk 2]
```diff
@@ -447,6 +474,129 @@ unsafe fn setup() -> (
...
+    );
+    kernel::deferred_call::DeferredCallClient::register(sip_hash);
+
+    // TicKV
+    let tickv = components::tickv::TicKVComponent::new(
+        sip_hash,
+        mux_flash,                                             // Flash controller
+        (0x100000 - 0x28000) / apollo3::flash_ctrl::PAGE_SIZE, // Region offset (Last 0x28000 bytes of flash)
+        // Region Size, the final page doens't work correctly
+        0x28000 - apollo3::flash_ctrl::PAGE_SIZE,
```
Comment (lschuermann) [inline boards/apollo3/lora_things_plus/src/main.rs:504]: This seems tied to the space that we reserve in the linker script. Can you make this a dedicated `MEMORY` region in the linker script and export start and end symbols to it? That would make it more obvious what those numbers mean, and should be more reliable in the face of changes to the linker script.

**Review by lschuermann — APPROVED** [review]
This looks good to me generally.

File: chips/apollo3/src/flash_ctrl.rs [hunk 3]
```diff
@@ -0,0 +1,350 @@
...
+/// @param addr - The location to be read.
+///
+/// @return The value read from the given address.
+unsafe fn flash_util_read_word(addr: *mut u32) -> u32 {
+    // Call `uint32_t flash_util_read_word(uint32_t *)` in the ROM code.
+    let flash_util_read_word: unsafe extern "C" fn(*mut u32) -> u32 =
+        unsafe { core::mem::transmute(0x08000075 as *const ()) };
+
+    flash_util_read_word(addr)
+}
```
Comment (bradjc) [inline chips/apollo3/src/flash_ctrl.rs:73]: Why are these functions marked unsafe? Isn't the unsafe operation mapping the function?

I'm not sure these can/should be marked safe, but I think that would be ideal as it isn't clear to me what the caller can do to ensure the operation is safe.

File: chips/apollo3/src/flash_ctrl.rs
Comment (bradjc) [inline chips/apollo3/src/flash_ctrl.rs:1]: Can this file be `flashcrtl.rs` like the others?

File: chips/apollo3/src/flash_ctrl.rs [hunk 4]
```diff
@@ -0,0 +1,350 @@
...
+/// This function will call the chip ROM code to perform the below operation.
+///
+/// This function will program multiple words in main flash.
+///
+/// @param program_key - The programming key, AM_HAL_FLASH_PROGRAM_KEY.
+/// @param src_addr - Pointer to word aligned array of data to program into
+/// the flash instance.
+/// @param dst_addr - Pointer to the word aligned flash location where
+/// programming of the flash instance is to begin.
+/// @param num_words - The number of words to be programmed.
```
Comment (bradjc) [inline chips/apollo3/src/flash_ctrl.rs:86]: This documentation isn't rendered anyway, but the `@param` makes it hard to read.

**Review by bradjc — COMMENTED** [review]
Looks good, just a couple comments. Mostly I'm unclear about the helper functions being marked `unsafe`.

File: chips/apollo3/src/flash_ctrl.rs (hunk 3 above)
Comment (lschuermann) [reply chips/apollo3/src/flash_ctrl.rs:73]: They can be used to read / write arbitrary flash memory, and thus arbitrary program code. As such, they are unsafe operations -- in much the same spirit that `mmap` is unsafe in Rust.

File: chips/apollo3/src/flash_ctrl.rs (hunk 4 above)
Comment (lschuermann) [reply chips/apollo3/src/flash_ctrl.rs:86]: Agreed, a simple markdown list is more idiomatic with Rust.

File: chips/apollo3/src/flash_ctrl.rs (hunk 3 above)
Comment (bradjc) [reply chips/apollo3/src/flash_ctrl.rs:73]: Got it.

Seems like `read_word()` is fine (ie safe).

The others should just copy what you wrote into the comment.

Comment (alistair23) [reply chips/apollo3/src/flash_ctrl.rs:73]: We don't know that though, so I do think it makes sense to keep them all unsafe

**Review by alevy — APPROVED** [review]

File: chips/apollo3/src/flash_ctrl.rs (hunk 3 above)
Comment (alevy) [reply chips/apollo3/src/flash_ctrl.rs]: > Seems like read_word() is fine (ie safe).

I think it should be unsafe as well.

This allows any capsule (with access to the method which, admittadly, it is private now) to read any flash data, which, in general, violates memory safety. It's only safe in the specific context that the `addr` points to a valid address that the caller is allowed to read, as  `u32`.
